File: Nim/K03.nim
import os
import random
import times
from strutils import parseInt
import IExecutable

type K03* = ref object of IExecutable

method question1*(this: K03) =
    stdout.write "年齢を入力してください。＞"
    stdout.flushFile
    var age = parseInt(readLine(stdin))
    if age < 20:
        echo "未成年なので購入できません。"

method question2*(this: K03) =
    stdout.write "身長を入力してください。＞"
    stdout.flushFile
    var height = parseInt(readLine(stdin)).toFloat
    height /= 100.0
    stdout.write "体重を入力してください。＞"
    stdout.flushFile
    var weight = parseInt(readLine(stdin)).toFloat
    var standard = height * height * 22.0
    echo "あなたの標準体重は", standard, "です。"

    if weight > standard and (weight - standard) / standard * 100.0 > 14.0:
        echo "太り気味です。"
    elif weight < standard and (weight - standard) / standard * 100.0 < -14.0:
        echo "痩せ気味です。"
    else:
        echo "普通ですね。"

method question3*(this: K03) =
    randomize(getTime().toWinTime)
    var number = rand(99)
    echo "０から９９の範囲の数値が決定されました。"
    stdout.write "決められた数値を予想し、この数値よりも大きな値を入力してください＞"
    stdout.flushFile
    var guess = parseInt(readLine(stdin))
    echo "決められた数値は", number, "です。"

    if guess > number:
        echo "正解です。"
    else:
        echo "不正解です。"

method question4*(this: K03) =
    randomize(getTime().toWinTime)
    var number = rand(99)
    echo "０から９９の範囲の数値が決定されました。"
    stdout.write "決められた数値を予想し、この数値よりも大きな値を入力してください＞"
    stdout.flushFile
    var guess = parseInt(readLine(stdin))
    echo "決められた数値は", number, "です。"

    if guess < 0 or guess > 100:
        echo "反則です！"
    elif guess > number and guess - number <= 10:
        echo "大正解です！"
    elif guess < number and guess - number >= -10:
        echo "惜しい！"
    elif guess == number:
        echo "お見事！"
    else:
        if guess > number:
            echo "正解です。"
        else:
            echo "不正解です。"

File: Nodejs/K03.ts
import { IExecutable } from './IExecutable';
import * as readline from 'readline';

export class K03 extends IExecutable {

    Question1(rl?: readline.ReadLine): Promise<string> {
        try {

            let age: number;

            let question1 = () => {
                return new Promise((resolve, reject) => {
                    rl.question("年齢を入力して下さい＞", (value) => {
                        age = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise(async (resolve, reject) => {
                    await question1();
                    if (age < 20) {
                        console.log("未成年なので購入できません。");
                    }
                    resolve();
                });
            };

            return new Promise(async (resolve, reject) => {
                await showResult();
                resolve();
            });

        } catch (e) {
            console.error(e);
        }
    }

    Question2(rl?: readline.ReadLine): Promise<string> {
        try {

            let height: number;
            let weight: number;

            let question1 = () => {
                return new Promise((resolve, reject) => {
                    rl.question("身長を入力して下さい＞", (value) => {
                        height = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let question2 = () => {
                return new Promise((resolve, reject) => {
                    rl.question("体重を入力して下さい＞", (value) => {
                        weight = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise(async (resolve, reject) => {
                    await question1();
                    await question2();
                    height /= 100.0;
                    let standard = height * height * 22.0;
                    console.log("あなたの標準体重は" + standard + "です。");

                    if (weight > standard && (weight - standard) / standard * 100.0 > 14.0) {
                        console.log("太り気味です。");
                    }
                    else if (weight < standard && (weight - standard) / standard * 100.0 < -14.0) {
                        console.log("痩せ気味です。");
                    }
                    else {
                        console.log("普通ですね。");
                    }
                    resolve();
                });
            };

            return new Promise(async (resolve, reject) => {
                await showResult();
                resolve();
            });

        } catch (e) {
            console.error(e);
        }
    }

    Question3(rl?: readline.ReadLine): Promise<string> {
        try {
            let randNum = Math.round(Math.random() * 100);

            console.log("０から９９の範囲の数値が決定されました。");

            let guess: number;
            let q1 = () => {
                return new Promise((resolve, reject) => {
                    rl.question("決められた数値を予想し、この数値よりも大きな値を入力してください＞", (value) => {
                        guess = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise(async (resolve, reject) => {
                    await q1();
                    console.log("決められた数値は" + randNum + "です。");
                    console.log(guess > randNum ? "正解です。" : "不正解です。");
                    resolve();
                });
            };

            return new Promise(async (resolve, reject) => {
                await showResult();
                resolve();
            });
        } catch (e) {
            console.error(e);
        }
    }

    Question4(rl?: readline.ReadLine): Promise<string> {
        try {
            let randNum = Math.round(Math.random() * 100);

            console.log("０から９９の範囲の数値が決定されました。");

            let guess: number;
            let q1 = () => {
                return new Promise((resolve, reject) => {
                    rl.question("決められた数値を予想し、この数値よりも大きな値を入力してください＞", (value) => {
                        guess = Number.parseInt(value);
                        resolve();
                    });
                });
            };

            let showResult = () => {
                return new Promise(async (resolve, reject) => {
                    await q1();
                    console.log("決められた数値は" + randNum + "です。");

                    if (guess > 100 || guess < 0) {
                        console.log("反則です！");
                    } else if (guess > randNum && guess - randNum <= 10) {
                        console.log("大正解です！");
                    } else if (guess < randNum && guess - randNum >= -10) {
                        console.log("惜しい！");
                    } else if (guess == randNum) {
                        console.log("お見事！");
                    } else {
                        console.log(guess > randNum ? "正解です。" : "不正解です。");
                    }

                    resolve();
                });
            };

            return new Promise(async (resolve, reject) => {
                await showResult();
                resolve();
            });
        } catch (e) {
            console.error(e);
        }
    }

}